ToolChain/Assembler/Assembler.py

```diff
@@ -56,79 +58,109 @@ def __init__(self, inputFile=None, outputFile=None):
-        for reference in self.parser.referenceDict:
-            if EXPORTED_REFERENCE_INDICATOR not in reference:
-                validInternalRef.append(reference)
-                fileContent += b"<refName>" + reference.encode("utf-8") + b"</refName>"
-                fileContent += b"<refAdd>" + struct.pack(">I", self.parser.referenceDict[reference]) + b"</refAdd>"
-
-                # Update global offsets would it be required
-                # This is required since all global offsets are set to 0 at the moment they are parsed
-                wouldBeExported = EXPORTED_REFERENCE_INDICATOR + " " + reference
-                if wouldBeExported in self.parser.referenceDict.keys():
-                    self.parser.referenceDict[wouldBeExported] = self.parser.referenceDict[reference]
-
-        fileContent += b"</InternalSymbols>"
-
-        # External symbols allow for a file to be linked with another file
-        fileContent += b"<ExternalSymbols>"
-        for reference in self.parser.referenceDict:
-            if EXPORTED_REFERENCE_INDICATOR in reference:
-                isValid = False
-                shortReference = reference.split()[1]
-                for internalReference in validInternalRef:
-                    if shortReference == internalReference:
-                        isValid = True
-                        break
-
-                if isValid:
-                    refName = reference.split()[1]
-                    fileContent += b"<refName>" + refName.encode("utf-8") + b"</refName>"
-                    fileContent += b"<refAdd>" + struct.pack(">I", self.parser.referenceDict[reference]) + b"</refAdd>"
+
+        parser = Parser()                    # Parser object which will return the binary code for instructions
+        binaryDataString = b"<Text>"         # String that will contain our binary code
+        xmlString = b""                      # String that will contain the "xml" data at the beginning of the .o file
+        globalSymbols = []                   # Temporary list to hold the names of global(external) symbols
+        localSymbols = []                    # Temporary list to hold the names of local(internal) symbols
+        externalSymbols = []                 # Master list of all external symbols to be used by linker
+        internalSymbols = []                 # Master list of all internal symbols
+        offset = 0                           # Holds the instruction's length, used to calculate offset
+        lineno = 0                           # Keeps track of which line we're parsing to pinpoint errors
+        relativeAddressCounter = 0           # Used to determine memory address of a label at a given point in the file
+
+        try:
+            # Initial test to verify that the file we're reading actually exists
+            file = open(inputFile, mode="r")
+            fileLines = file.readlines()
+            file.close()
+        except KeyError as e:
```

Comment:
KeyError is not raised by open. Please refer to: https://docs.python.org/3.5/library/functions.html#open
Also, close, can potentially cause a ValueError.

Reply:
It would be better to accept a general exception here since this included error message from sys will be sufficient to explain that opening the file failed. 

Alternatively, the file can be opened and read before reaching the assembler, aka in the run script in the root directory of the project.